beluga_ros/include/beluga_ros/amcl.hpp

```diff
@@ -45,33 +58,32 @@ struct AmclParams {
 };
 
 /// Implementation of the Adaptive Monte Carlo Localization (AMCL) algorithm.
+/**
+ * See Probabilistic Robotics \cite thrun2005probabilistic Chapter 6.2.
+ *
+ * \tparam SensorModel Type representing a sensor model. Must satisfy \ref SensorModelPage.
+ **/
+template <typename SensorModel>
 class Amcl {
  public:
   using particle_type = std::tuple<Sophus::SE2d, beluga::Weight>;
 
-  using motion_model_variant = std::variant<
-      beluga::DifferentialDriveModel,     //
-      beluga::OmnidirectionalDriveModel,  //
-      beluga::StationaryModel>;
+  using sensor_model_type = SensorModel;
 
-  using sensor_model_variant = std::variant<
-      beluga::LikelihoodFieldModel<beluga_ros::OccupancyGrid>,  //
-      beluga::BeamSensorModel<beluga_ros::OccupancyGrid>>;
-
-  using execution_policy_variant = std::variant<std::execution::sequenced_policy, std::execution::parallel_policy>;
+  using map_type = typename sensor_model_type::map_type;
 
   /// Constructor.
   /**
-   * \param map Occupancy grid map.
+   * \param map Map suited for the sensor model.
    * \param motion_model Variant of motion model.
-   * \param sensor_model Variant of sensor model.
+   * \param sensor_model An instance of the sensor model.
    * \param params Parameters for AMCL implementation.
    * \param execution_policy Variant of execution policy.
    */
   Amcl(
-      beluga_ros::OccupancyGrid map,
+      map_type map,
       motion_model_variant motion_model,
-      sensor_model_variant sensor_model,
+      sensor_model_type sensor_model,
```

Comment:
@serraramiro1 meta: I get the idea of bubbling up variants to calling code and keeping this class fully statically typed instead. What I'm not very fond of is the mix of having some of it fully statically typed (i.e. the sensor model) and some of it using algebraic data types (i.e. the motion model). IMO it should be either or but not both.
Of course, if you go fully statically typed you will have a combinatorial explosion in your `AnyAMCL` variant definition. 

All that to say that I'm not entirely convinced this is the right path (but I can be swayed too). Have you considered introducing a "filter initializer" object instead of passing the map to the filter constructor? Have you considered templating the variant types themselves? If you are worried about enforcing constraints between types, that can be arranged.

Reply:
You're totally right.
Based on the offline sync we had:

- I'll create a new fully typed generic pipeline that can live in `beluga`
- Create a ticket to consolidate the `beluga_ros` pipeline with the new one in beluga